src/main/java/org/opentripplanner/standalone/server/EtagRequestFilter.java

```diff
@@ -0,0 +1,75 @@
+package org.opentripplanner.standalone.server;
+
+import com.google.common.hash.Hashing;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import javax.ws.rs.HttpMethod;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerResponseContext;
+import javax.ws.rs.container.ContainerResponseFilter;
+import org.apache.commons.codec.binary.Hex;
+import org.apache.http.HttpStatus;
+import org.opentripplanner.ext.vectortiles.VectorTilesResource;
+
+public class EtagRequestFilter implements ContainerResponseFilter {
+
+  public static final String DIRECTIVE_NO_STORE = "no-store";
+  public static final String HEADER_ETAG = "ETag";
+  public static final String HEADER_IF_NONE_MATCH = "If-None-Match";
+  public static final String HEADER_CONTENT_TYPE = "Content-Type";
+
+  @Override
+  public void filter(ContainerRequestContext request, ContainerResponseContext response)
+    throws IOException {
+    if (
+      isEligibleForEtag(request, response) &&
+      hasAllowedContentType(response) &&
+      response.getEntity() instanceof byte[] bytes &&
+      bytes.length > 0
+    ) {
+      var clientEtag = request.getHeaderString(HEADER_IF_NONE_MATCH);
+      var etag = generateETagHeaderValue(bytes);
+      var headers = response.getHeaders();
+      headers.add(HEADER_ETAG, etag);
```

Comment:
According to your screenshot, OTP currently doesn't support `Accept-Encoding`/`Content-Encoding`; But once it ever does, setting `Vary` will be necessary for caching to work as intended. The same applies to `Accept`/`Content-Type` and `Accept-Language`/`Content-Language`. Since setting `Vary` doesn't hurt, I recommend to just do it now:

```suggestion
      headers.add("Vary", "Accept, Accept-Encoding, Accept-Language");
      headers.add(HEADER_ETAG, etag);
```

Reply:
I think this should probably be done in a separate filter, and for all responses?

Reply:
I added a separate filter for Vary in https://github.com/opentripplanner/OpenTripPlanner/pull/4511/commits/4ea0e35cb63a7b2f5cb2cf569cafa06f2be4240a